### ISSUE TITLE
[kore] default program start symbol

### ISSUE THREAD

**Issue by radumereuta** [kframework/k-legacy#1520]
In K3 we have the convention that the sort of $PGM from the configuration specifies the default start symbol for programs. I would like to have some sort of similar convention for the kore implementation also, because right now it defaults to K, and I have to specify the start symbol each time I run kast.

@dwightguth I've looked over the compilation process, but I'm not sure how to extract the information I need, so I'm assigning it to you. I can do it if you tell me how, but I think this is a really easy fix for you.

**Comment by dwightguth** [comment]
I believe that the way to go is to iterate over the body of every initializer rule in the parsed definition looking for a term of the following form:

#SemanticCastToPgm(Map:lookup(Init,#token("$PGM",KConfigVar)))

In this case, you can deduce that the start symbol of $PGM is Pgm. This is not immediately urgent enough for me in order for me to work on this myself, but feel free to ask me any questions if you run into issues implementing it.

**Comment by radumereuta** [comment]
Can I encounter multiple instances of $PGM?
Currently there is a default configuration in `kast.k`. Are you doing any processing to clean that up before getting to serialization to disk?

**Comment by dwightguth** [comment]
The default configuration only is actually imported in the definition if you either explicitly import it, or the definition contains no other configuration declaration. As long as you examine only the main module of the parsed definition (`compiledDefinition.getParsedDefinition().mainModule()`), you shouldn't have to worry about anything other than the actual initializer rules that you find in that module.

As far as multiple instances of the same variable go, you can encounter something like that, yes. I suggest that what you do if you encounter this behavior is compute the least upper bound of all the sorts that you see, use that value as the start symbol for the configuration variable, and emit a warning.

**Comment by radumereuta** [comment]
Hmmm... ok, this looks like a good start. I will get back to this problem when I make more progress with JS6.

**Comment by bmmoore** [comment]
Looks CompiledDefinition.programStartSymbol does save and restore a sort, but KompileFrontEnd.java:88 currently hardcodes it to K.

**Comment by bmmoore** [comment]
We should gather sorts for all configuration variables as part of fixing this.

**Comment by bmmoore** [comment]
@cos I think we need to extend the saved definition to record also a sort for all configuration variables to be able to solve this. You wanted to handle changes to those classes, right?

**Comment by cos** [comment]
I get the feeling we're again adding even more special casing to K with this automated configuration variable start symbol hardwiring. How about if the user wants to grab $PGM as a string and use a different parser? Why don't we simply write `#parse($PGM, Exp)` in all cases? Is it that verbose?

**Comment by cos** [comment]
I see `#parse` takes a file for some reason. We might do via something like `#parse(#readFile("filename"), Exp)`. Furthermore, why have `$PGM` as a special variable to begin with. Can't we simply hook it to a function reading configuration variables instead of adding yet another feature? Something like `#config("PGM")`? Or, fi we do allow it to be `$PGM`, maybe implement it as a hooked function instead of hardwiring: 

```
syntax K ::= "$" r"[a-z0-9] or whatever regex or other sort" [ hook(readVariable) ]
```

This way, the entire compilation pipeline will be oblivious to configuration variables. Only the runtime will have to provide the appropriate function and hook it to the input config Map.

**Comment by bmmoore** [comment]
That sounds like a lot more work than finishing the implementation of the approach we've been using since K 2.

**Comment by cos** [comment]
It should be fairly simple if `#parse` is already working. Is it?
If it is, it simply means replacing the code that tried to parse the `$PGM` with code that simply passes everything in as a String.

**Comment by dwightguth** [comment]
Pretty sure this causes issues for the OCAML backend because #parse is
quite nontrivial to implement there. I would prefer to avoid the giant
headache and loss of productivity that would result in trying to deal with
this. Not to mention the loss of performance.

**Comment by cos** [comment]
for the OCAML backend, if we propagate the $PGM value you can simply replace `#parse(theValue, theSort)` with the actual parse, just like it was done before, but without having to worry about recovering the sort.

In any case, as long as we're not saving the $PGM sort in `Module` (`CompiledDefinition` is ok as it is already a bit messy) and I'm not the one implementing this, I'm ok with any solution.

**Comment by bmmoore** [comment]
Making or approving `CompiledDefinition` changes is the main reason I assigned the bug to you. The current code already serializes and deserializes a sort for `$PGM` and passes it to the parser! The problem is only that we don't have a field for storing similarly the sorts of configuration variables, and that the construction of the compiled definition always passes `K` as the sort to be saved instead of computing it properly.

**Comment by cos** [comment]
Let's try to fix this. 
@radumereuta, why aren't configuration variables true variables?

**Comment by cos** [comment]
fixed by #2093
